homeassistant/components/media_player/dlna_dmr.py

```diff
@@ -0,0 +1,909 @@
+        state_var.value = result['CurrentTransportState']
+
+        if old_value != result['CurrentTransportState']:
+            self.on_state_variable_change(service, [state_var])
+
+        return self.state
+
+    @asyncio.coroutine
+    def _async_poll_position_info(self, action):
+        """Update position info."""
+        result = yield from action.async_call(InstanceID=0)
+
+        service = action.service
+        track_duration = service.state_variable('CurrentTrackDuration')
+        track_duration.value = result['TrackDuration']
+
+        time_position = service.state_variable('RelativeTimePosition')
+        time_position.value = result['RelTime']
+
+        self.on_state_variable_change(service, [track_duration, time_position])
+
+    def on_state_variable_change(self, service, state_variables):
+        """State variable(s) changed, let home-assistant know."""
+        self.schedule_update_ha_state()
+
+    @property
+    def supported_features(self):
+        """Flag media player features that are supported."""
+        supported_features = 0
+
+        if not self._device:
+            return supported_features
+
+        rc_service = self._service('RC')
+        if rc_service:
+            if rc_service.state_variable('Mute'):
+                supported_features |= SUPPORT_VOLUME_MUTE
+            if rc_service.state_variable('Volume'):
+                supported_features |= SUPPORT_VOLUME_SET
+
+        avt_service = self._service('AVT')
+        if avt_service:
+            state_var = avt_service.state_variable('CurrentTransportActions')
+            if state_var:
+                value = state_var.value or ''
+                actions = value.split(',')
+                if 'Play' in actions:
+                    supported_features |= SUPPORT_PLAY
+                if 'Stop' in actions:
+                    supported_features |= SUPPORT_STOP
+                if 'Pause' in actions:
+                    supported_features |= SUPPORT_PAUSE
+
+            current_track_var = avt_service.state_variable('CurrentTrack')
+            num_tracks_var = avt_service.state_variable('NumberOfTracks')
+            if current_track_var and \
+               num_tracks_var and \
+               current_track_var.value is not None and \
+               num_tracks_var.value is not None:
+                current_track = current_track_var.value
+                num_tracks = num_tracks_var.value
+                if current_track > 1:
+                    supported_features |= SUPPORT_PREVIOUS_TRACK
+
+                if num_tracks > current_track:
+                    supported_features |= SUPPORT_NEXT_TRACK
+
+            play_media_action = avt_service.action('SetAVTransportURI')
+            play_action = avt_service.action('Play')
+            if play_media_action and play_action:
+                supported_features |= SUPPORT_PLAY_MEDIA
+
+        return supported_features
+
+    @property
+    @requires_state_variable('RC', 'Volume')
```

Comment:
Properties must not execute any IO

All property values must be populated in async_update(), if I understand correctly requires_state_variable will trigger some UPnP message exchange

Reply:
This does not do any IO. In `DlnaDmrDevice._async_init_device()` the specification of the device is parsed. Via the `NotifyView` the any values for state variables are stored. These methods simply read whats already stored.